### PR TITLE
add `Compilation` API

### PR DESCRIPTION
**Pull request by OmarTawfik**
And along, a few fixes to the parser and binding APIs (more details in the added changesets).

### CODE REVIEW

File: crates/codegen/runtime/cargo/wasm/src/runtime/wrappers/bindings/mod.rs
Comment (OmarTawfik) [inline crates/codegen/runtime/cargo/wasm/src/runtime/wrappers/bindings/mod.rs:18]: @ggiraldez these are the wrappers I mentioned, due to the fact that the original `Definition` and `Reference` types have handles in them.
They should be removed once the `metaslang_bindings` API is fixed.

File: crates/metaslang/bindings/src/lib.rs
Comment (OmarTawfik) [inline crates/metaslang/bindings/src/lib.rs:271]: @ggiraldez unrelated to the PR, but I wonder if this `TODO` is already fixed?

> // TODO: what should we do if the parent reference
> // cannot be resolved at this point?

File: crates/codegen/runtime/cargo/wasm/src/runtime/wrappers/bindings/mod.rs
Comment (ggiraldez) [reply crates/codegen/runtime/cargo/wasm/src/runtime/wrappers/bindings/mod.rs:18]: Is the problem we don't want to expose those handles to Typescript? The handles themselves are `usize` values. Or is the pointer/reference to the bindings object that's the problem? I recall you mentioning that they should be copyable.

Since we're going to pre-resolve all references anyway, we may just get away with cursors in the end. But asking to understand and to be aware of what are the limitations.

File: crates/metaslang/bindings/src/lib.rs
Comment (ggiraldez) [reply crates/metaslang/bindings/src/lib.rs:271]: It's not, but I don't know if there's something meaningful we can do in this case. For now, if we cannot resolve a parent it's simply ignored.

This resolution method is used internally, while resolving other references. Eg. for determining which methods are reachable when resolving a virtual call; or to determine which extension scopes (ie. attached functions) apply in a particular context. Resolving parents in these circumstances should not be a problem unless the user code is invalid, or if there's a bug in the binding rules.

File: crates/solidity/outputs/npm/tests/src/compilation/binding-graph.test.mts
Comment (ggiraldez) [reply crates/solidity/outputs/npm/tests/src/compilation/binding-graph.test.mts:103]: Correct. This is probably because there's a missing step when constructing the `BindingsGraph` in `CompilationUnit` to add the built-ins. Since this is language dependant, I don't think it's possible to do in the runtime. As of today's base `main` this is also a bit cumbersome, but I streamlined the process in #1170. It may be worth it to wait for it to be merged (shouldn't be too far ahead) to find a good solution for this.

File: crates/solidity/testing/perf/src/tests/init_bindings.rs
Comment (ggiraldez) [inline crates/solidity/testing/perf/src/tests/init_bindings.rs:32]: I don't think this is important because the files being tested are unrelated, but for correctness this should remove the quotes surrounding the string literal received.

File: crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs
Comment (ggiraldez) [reply crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs:12]: After reading the whole PR I got to the actual `CompilationBuilder`. I still find the `Internal` prefix a bit confusing here, but it's more a nit than anything else.

File: crates/codegen/runtime/cargo/crate/src/runtime/compilation/file.rs
Comment (OmarTawfik) [reply crates/codegen/runtime/cargo/crate/src/runtime/compilation/file.rs:1]: good catch!

File: crates/codegen/runtime/cargo/wasm/src/runtime/wrappers/bindings/mod.rs
Comment (OmarTawfik) [reply crates/codegen/runtime/cargo/wasm/src/runtime/wrappers/bindings/mod.rs:18]: > is the pointer/reference to the bindings object that's the problem? 

Yes. For WIT resources, we need to track their lifetimes on the other side of the FFI boundary, and there isn't a straightforward way to do that with the handle references.
For the existing WIT resources, we either wrap them in `Rc<>` (`define_rc_wrapper()`) or `RefCell<>` (`define_refcell_wrapper()`).

File: crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs
Comment (OmarTawfik) [reply crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs:12]: I was thinking that we would then expose a similar callback-based API in Rust as well using a new `CompilationBuilder` that would wrap the `InternalCompilationBuilder` here. WDYT?
I'm happy to rename it as well if you have a better idea.

File: crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs
Comment (OmarTawfik) [reply crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs:49]: Sure. We do that check in TS, but wouldn't hurt to add it here as well.

File: crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs
Comment (OmarTawfik) [reply crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs:107]: The difference is that unparsing a non-terminal can also return the trivia around it, so it will be confusing for users to deal with leading whitespace in `  "foo.sol"`.

File: crates/metaslang/bindings/src/lib.rs
Comment (OmarTawfik) [reply crates/metaslang/bindings/src/lib.rs:271]: > unless the user code is invalid

Which is a possible use case, right?

> if we cannot resolve a parent it's simply ignored

Sounds good. I will use `None` in this case.

File: crates/solidity/outputs/cargo/tests/src/bindings.rs
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/tests/src/bindings.rs:27]: Will move it to the extensions module.

File: crates/codegen/runtime/cargo/crate/src/runtime/compilation/unit.rs
Comment (OmarTawfik) [reply crates/codegen/runtime/cargo/crate/src/runtime/compilation/unit.rs:47]: Good catch. Will move to extensions.

File: crates/metaslang/bindings/src/lib.rs
Comment (ggiraldez) [reply crates/metaslang/bindings/src/lib.rs:271]: Yes, it's certainly possible that the user code is invalid and that means we cannot resolve a parent reference. Eg. definining `contract Test is Base {}` without defining `Base`. This failure to resolve is currently ignored. I think my main concern for writing the `TODO` was that while this may impact resolved references, there's no feedback given to the user indicating that this may be a problem. In general, the current behaviour should be ok though.

File: crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs
Comment (ggiraldez) [reply crates/codegen/runtime/cargo/crate/src/runtime/compilation/internal_builder.rs:107]: Ah! I see. Thanks for the explanation.

File: crates/solidity/outputs/npm/tests/src/compilation/binding-graph.test.mts
Comment (OmarTawfik) [reply crates/solidity/outputs/npm/tests/src/compilation/binding-graph.test.mts:103]: I made the refactoring, and added a couple of `TODO`s that you might want to check for the remaining work. I didn't want to do a larger refactoring of bindings until you get a chance to review/rebase on top.

File: crates/solidity/inputs/language/src/definition.rs
Comment (ggiraldez) [inline crates/solidity/inputs/language/src/definition.rs:7]: I was thinking it may make sense to put this in the new `LanguageFacts` struct. WDYT?

**Review by ggiraldez — APPROVED** [review]
LGTM! Thanks for making some of the changes we've discussed in #1170. I'll rebase that once this is merged.

File: crates/solidity/inputs/language/src/definition.rs
Comment (OmarTawfik) [reply crates/solidity/inputs/language/src/definition.rs:7]: mmm, What is the use case to expose this in the public API?

**Pull request merged**
